### PR TITLE
Model and Test

### PR DESCRIPTION
**Pull request by natalie2nguyen**
Created a working model trained on OIT questions that predicts the intent. Tests written as well.

### CODE REVIEW

File: server/requirements.txt [hunk 1]
```diff
@@ -10,13 +11,25 @@ Jinja2==3.1.4
 joblib==1.4.2
 MarkupSafe==2.1.5
 nltk==3.9.1
-numpy==2.1.1
+numpy==1.26.4
```
Comment (Ry-ka) [inline server/requirements.txt]: Can you upgrade this to 2.x

File: server/src/model.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+# clean and preprocess the questions (remove punctuation, lowercase the text, etc.).
+# remove stop words like "is", "my", "the"
+# use TF-IDF (Term Frequency-Inverse Document Frequency) to convert the questions into numerical vectors
+# in comparison to CountVectorizer, CV just counts the frequency of a word in a given sentence/input
+# TF_IDF counts the frequency AND the Inverse Document Frequency
+# use Random Forest Classifier 
+# predict the intent of the user's question/input
+
+import re
+def model(question: string):
```
Comment (Ry-ka) [inline server/src/model.py]: Can we split the training and the output so that we're not training the model everytime we need the output? This is inefficient. Consider saving the model into a joblib like we did in class.

File: server/src/model.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+import string
+import pandas as pd
+from sklearn.model_selection import train_test_split
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.metrics import accuracy_score, classification_report
```
Comment (Ry-ka) [inline server/src/model.py]: I don't see this being used? Unless I'm mistaken.

File: server/src/model.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+# use Random Forest Classifier 
+# predict the intent of the user's question/input
+
+import re
+def model(question: string): 
+    df = pd.read_excel('server/src/dataset/OIT_Dataset.xlsx', sheet_name="Sheet1")
+    df = df.sample(frac=1)
+
+    df['Question'] = df['Question'].str.lower().str.replace('[^\w\s]', '', regex=True)
+    # nltk.download('stopwords')
```
Comment (Ry-ka) [inline server/src/model.py]: Probably call this outside the function so that we don't attempt to download this everytime.

File: server/src/model.py (hunk 2 above)
Comment (Ry-ka) [reply server/src/model.py]: Right now, the function does preprocessing, training, and output all in one function. Let's clean up the comments and split the function into 3 (actually 2 because we already have a pre-processing function).

File: server/tests/test_model.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    'Question': ['how do i reset my password', 'what is my username'],
+    'Intent': ['Password_reset', 'Username']
+})
+
+class TestIntentModel(unittest.TestCase):
+
+    @patch('model.pd.read_excel')
+    @patch('model.TfidfVectorizer')
+    @patch('model.RandomForestClassifier')
+    def test_question(self, mock_rf, mock_vectorizer, mock_read_excel):
```
Comment (Ry-ka) [inline server/tests/test_model.py]: This test isn't atomic enough.

Comment (Ry-ka) [reply server/tests/test_model.py]: Split this up into multiple tests after splitting your logic first.

File: server/tests/test_model.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        # Mocking stopwords and dataset
+        mock_read_excel.return_value = mocked_data
+        
+        # Mock the TfidfVectorizer and RandomForestClassifier behavior
+        mock_vectorizer_instance = MagicMock()
+        mock_vectorizer_instance.transform.return_value = MagicMock()
+        mock_vectorizer_instance.fit_transform.return_value = MagicMock()
+        mock_vectorizer.return_value = mock_vectorizer_instance
```
Comment (Ry-ka) [inline server/tests/test_model.py]: You're over mocking, and don't actually test any logic here.

**Review by Ry-ka — CHANGES_REQUESTED** [review]
Please see comments

File: server/src/model.py (hunk 4 above)
Comment (natalie2nguyen) [reply server/src/model.py]: which function?

File: server/tests/test_model.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        # Test the predict_intent function with a sample question
+        for i, input_question in enumerate(mocked_data['Question']): 
+            predicted_intent = predict_intent(trained_model, trained_vectorizer, input_question)
+
+            # Increment count if prediction is correct
+            if predicted_intent == mocked_data['Intent'][i]:
+                count += 1
+
+        # Check if the count of correct predictions is greater than 10
+        assert count > 5, f"Test failed. Only {count} correct predictions."
```
Comment (Ry-ka) [inline server/tests/test_model.py]: The comment says greater than 10, but the test checks for greater than 5. Also, why 10? Wouldn't it be better to use accuracy or an error metric that we learned about in class?

**Review by Ry-ka — APPROVED** [review]
See comment for question.

**Review by ErnestoRiDo — APPROVED** [review]
